### ISSUE TITLE
Interesting behaviour when applying list transform to the result of an aggregation

### ISSUE THREAD

**Issue by Avalon8D** [duckdb/duckdb#9525]
### What happens?

This is the simplest query I could devise:
```sql
select
    list_transform(
        first(example),
        _dummy -> _dummy + 1
    )
from
    (select range(2) as example from range(10))
group by
    example
```
Error:
```
BinderException                           Traceback (most recent call last)
Cell In[3], line 1
----> 1 duckdb.sql("""\
      2 select
      3     list_transform(
      4         first(list_of_arrays.example),
      5         _dummy -> _dummy + 1
      6     )
      7 from
      8     (select range(2) as example from range(10)) as list_of_arrays
      9 group by
     10     example\
     11 """)

BinderException: Binder Error: column "_dummy" must appear in the GROUP BY clause or must be part of an aggregate function.
Either add it to the GROUP BY list, or use "ANY_VALUE(_dummy)" if the exact value of "_dummy" is not important.
```

Basically, the `_dummy` name is recognized as a column that needs to be aggregated. 
My understanding would be that the list transform is an operation applied to the result of the group by.
That said, there is also the scenario where you put columns that have not been aggregated into the lambda expression.
This can be less of a bug and more of a 'what to expect' sort of situation.

I tried the same with correlate a correlate subquery. Not sure if the behaviours should match, but thought it would be useful to compare the behaviours:
```sql
select
    (
        select
            list_transform(
                an_array,
                _dummy -> _dummy + 1
            )
        from
            (select unnest(first(example))) as an_array
    ) as first_array_plus_1
from
    (select range(2) as example from range(10)) as list_of_arrays
group by
    example
```
Error:
```
---------------------------------------------------------------------------
BinderException                           Traceback (most recent call last)
Cell In[4], line 1
----> 1 duckdb.sql("""\
      2 select
      3     (
      4         select
      5             list_transform(
      6                 an_array,
      7                 _dummy -> _dummy + 1
      8             )
      9         from
     10             (select unnest(first(example))) as an_array
     11     ) as first_array_plus_1
     12 from
     13     (select range(2) as example from range(10)) as list_of_arrays
     14 group by
     15     example\
     16 """)

BinderException: Binder Error: Referenced column "_dummy" not found in FROM clause!
Candidate bindings: "an_array.unnest(first(example))"
```

And we get a completely different error! Though it also alludes to the same lambda name binding situation.

I did not find any caveats listed in the docs about this:
- https://duckdb.org/docs/sql/functions/nested#lambda-functions
- https://duckdb.org/docs/sql/expressions/subqueries#correlated-subqueries

Hope I provided enough resources :)

### To Reproduce

```python
## !pip install duckdb -U --pre # if installing in ipython or jupyter

import duckdb

# list transform
duckdb.sql("""\
select
    list_transform(
        first(list_of_arrays.example),
        (_dummy) -> _dummy + 1
    ) as first_array_plus_1
from
    (select range(2) as example from range(10)) as list_of_arrays
group by
    example\
""")

# correlate subquery
duckdb.sql("""\
select
    (
        select
            list_transform(
                an_array,
                (_dummy) -> _dummy + 1
            )
        from
            (select unnest(first(example))) as an_array
    ) as first_array_plus_1
from
    (select range(2) as example from range(10)) as list_of_arrays
group by
    example\
""")
```

### OS:

iOS

### DuckDB Version:

0.9.2.dev231,0.9.1

### DuckDB Client:

Python

### Full Name:

Pedro Schmidt

### Affiliation:

Booking.com

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by taniabogatsch** [comment]
Thanks for filing this and for the easily reproducible example.

I got curious about this issue as it was related to lambdas. This looks similar to fixes that I already addressed as part of this lambda binding refactor PR: https://github.com/duckdb/duckdb/pull/8851. See the referenced issue https://github.com/duckdb/duckdb/issues/5695 there.

We'll eventually merge the `feature` branch back into `main`. Sadly the changes were too involved to go directly into `main` as part of a bug fix.

I just pulled `feature` and it seems to work for the first query 🤞 
```sql
tania@motorbook duckdb % build/release/duckdb 
v0.9.2-dev579 5e9b28b8f2
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D select
>     list_transform(
>         first(example),
>         _dummy -> _dummy + 1
>     )
> from
>     (select range(2) as example from range(10))
> group by
>     example;
┌──────────────────────────────────────────────────────────┐
│ list_transform(first(example), (_dummy -> (_dummy + 1))) │
│                         int64[]                          │
├──────────────────────────────────────────────────────────┤
│ [1, 2]                                                   │
└──────────────────────────────────────────────────────────┘
```

The second query throws the expected error on `feature` (I think). It does no longer throw the misleading error message that you got on `main`.
```sql
D select unnest(first(example)) as an_array
> from
>     (select range(2) as example from range(10)) as list_of_arrays
> group by
>     example;
┌──────────┐
│ an_array │
│  int64   │
├──────────┤
│        0 │
│        1 │
└──────────┘
```
This returns a non-list input parameter for `list_transform`. This is my output:
```sql
D select
>     (
>         select
>             list_transform(
>                 an_array,
>                 _dummy -> _dummy + 1
>             )
>         from
>             (select unnest(first(example))) as an_array
>     ) as first_array_plus_1
> from
>     (select range(2) as example from range(10)) as list_of_arrays
> group by
>     example;
Error: Binder Error: failed to bind function, either: Invalid LIST argument during lambda function binding!
 or: Referenced column "_dummy" not found in FROM clause!
Candidate bindings: "an_array.unnest(first(example))"
```
> Invalid LIST argument during lambda function binding!

Seems expected!

**Comment by Avalon8D** [comment]
Hello Tania,

Happy to hear the issue has been addressed!

On the correlated subquery case, it does seem I'm referring to the wrong column there!
And on the correlate subquery, I do realise I wrote it wrong as well! And the new error message is helpful in identifying, after all it can be very easy to get lost on the nesting level!

Also a shout out to the list comprehension syntax! It's so much easier to read for these situations than the composition of the list_* operations!

I give the issue as closed :)
